Production/plugins/PixelTrackTableProducer.cc

```diff
@@ -0,0 +1,192 @@
+      int vtx_outIdx = -1;
+      if(iter != vtxGood.end())
+        vtx_outIdx = std::distance(vtxGood.begin(), iter);
+      vtxIdx[trk_outIdx] = vtx_outIdx;
+    }
+
+    auto table = std::make_unique<nanoaod::FlatTable>(nTrk, name, false, false);
+    table->addColumn<float>("pt", pt, "pt", precision_);
+    table->addColumn<float>("eta", eta, "eta", precision_);
+    table->addColumn<float>("phi", phi, "phi", precision_);
+    table->addColumn<float>("tip", tip, "tip", precision_);
+    table->addColumn<float>("zip", zip, "zip", precision_);
+    table->addColumn<float>("chi2", chi2, "chi2", precision_);
+    table->addColumn<float>("charge", charge, "charge", precision_);
+    table->addColumn<int>("quality", quality,
+      "track quality: bad = 0, edup = 1, dup = 2, loose = 3, strict = 4, tight = 5, highPurity = 6");
+    table->addColumn<int>("nLayers", nLayers, "number of layers with hits");
+    table->addColumn<int>("nHits", nHits, "number of hits");
+    table->addColumn<float>("dxy", dxy, "transverse IP wrt the beam spot", precision_);
+    table->addColumn<float>("dz", dz, "longitudinal IP wrt the beam spot", precision_);
+    table->addColumn<int>("vtxIdx", vtxIdx, "index of the associated vertex (-1 if none)");
+
+    event.put(std::move(table), name);
+  }
+
+  void FillVertices(edm::Event& event, const ZVertexSoA& vertices, const std::vector<int>& vtxGood) const
+  {
+    static const std::string name = "PixelVertex";
+    const size_t nVtx = vtxGood.size();
+
+    std::vector<float> zv(nVtx), wv(nVtx), chi2(nVtx), ptv2(nVtx);
+    std::vector<int> ndof(nVtx);
+
+    for(size_t vtx_outIdx = 0; vtx_outIdx < nVtx; ++vtx_outIdx) {
+      const int vtx_idx = vtxGood[vtx_outIdx];
+      zv[vtx_outIdx] = vertices.zv[vtx_idx];
+      wv[vtx_outIdx] = vertices.wv[vtx_idx];
+      chi2[vtx_outIdx] = vertices.chi2[vtx_idx];
+      ptv2[vtx_outIdx] = vertices.ptv2[vtx_idx];
+      ndof[vtx_outIdx] = vertices.ndof[vtx_idx];
+    }
+
+    auto table = std::make_unique<nanoaod::FlatTable>(nVtx, name, false, false);
+    table->addColumn<float>("z", zv, "z-position", precision_);
+    table->addColumn<float>("weight", wv, "weight (1/error^2)", precision_);
+    table->addColumn<float>("chi2", chi2, "chi2", precision_);
+    table->addColumn<float>("ptv2", ptv2, "pt^2", precision_);
+    table->addColumn<int>("ndof", ndof, "number of degrees of freedom");
+
+    event.put(std::move(table), name);
+  }
+
+  std::pair<std::vector<int>, std::vector<int>> selectGoodTracksAndVertices(const TrackSoA& tracks,
+                                                                            const ZVertexSoA& vertices) const
+  {
+    const uint32_t nv = vertices.nvFinal;
+    std::vector<int> trkGood, vtxGood;
+
+    std::vector<size_t> nTrkAssociated(nv, 0);
+    for(int32_t trk_idx = 0; trk_idx < tracks.stride(); ++trk_idx) {
+      const auto nHits = tracks.nHits(trk_idx);
+      if(nHits == 0) break;
```

Comment:
It should be `break`. We use on purpose the new low level gpu-friendly `TrackSoA` class without converting it to the legacy format, because such conversion is time costly. `TrackSoA` have a fixed size arrays and the convention is that the first entry with `nHits == 0` indicates the end of the track collection for the given event.